File: optimization-greedy-algorithms-problems-pro/src/main/java/com/greedy/problems/CoinChange.java
package com.greedy.problems;

import java.util.Arrays;

/**
 * **Problem: Coin Change Problem (Greedy Approach)**
 *
 * Given a set of coin denominations and a target amount, find the minimum number of coins
 * required to make up that amount.
 *
 * **Important Note on Greedy Strategy:**
 * The greedy approach (always choosing the largest denomination coin less than or equal to the remaining amount)
 * works *only* for certain "canonical" coin systems (e.g., US currency: {1, 5, 10, 25} cents, or most
 * common currencies like {1, 2, 5, 10, 20, 50} units).
 *
 * It **does NOT work** for all coin systems. For example, if denominations are {1, 3, 4} and target is 6:
 * - Greedy: 4 + 1 + 1 = 3 coins.
 * - Optimal: 3 + 3 = 2 coins.
 *
 * For arbitrary coin systems, the Dynamic Programming approach is required to guarantee optimality.
 * This problem focuses on demonstrating the greedy approach where it is applicable.
 *
 * **Greedy Strategy:**
 * 1. Sort the coin denominations in descending order.
 * 2. Iterate through the denominations, starting from the largest.
 * 3. For each denomination, take as many coins as possible without exceeding the remaining amount.
 * 4. Subtract the value of taken coins from the amount and repeat until the amount is zero.
 *
 * **Why it works for canonical systems (Intuition):**
 * Canonical systems are designed such that taking the largest possible coin at each step
 * never prevents an optimal solution from being formed. This property is usually proven
 * mathematically for specific systems, often by contradiction.
 */
public class CoinChange {

    /**
     * Solves the Coin Change Problem using a greedy approach.
     * This method assumes a canonical coin system where the greedy strategy is optimal.
     *
     * @param denominations An array of available coin denominations (e.g., {1, 5, 10, 25}).
     *                      It's assumed these are positive integers.
     * @param amount        The target amount to make change for. It's assumed to be non-negative.
     * @return The minimum number of coins required, or -1 if the amount cannot be made
     *         (though for canonical systems including 1, this shouldn't happen for non-negative amounts).
     */
    public int findMinCoinsGreedy(int[] denominations, int amount) {
        // Handle edge cases
        if (amount < 0) {
            throw new IllegalArgumentException("Amount cannot be negative.");
        }
        if (amount == 0) {
            return 0; // 0 coins needed for 0 amount
        }
        if (denominations == null || denominations.length == 0) {
            return -1; // Cannot make change without denominations
        }

        // Step 1: Sort denominations in descending order.
        // This ensures we always try to use the largest possible coin first.
        Arrays.sort(denominations);
        // Reverse the sorted array
        for (int i = 0; i < denominations.length / 2; i++) {
            int temp = denominations[i];
            denominations[i] = denominations[denominations.length - 1 - i];
            denominations[denominations.length - 1 - i] = temp;
        }

        int coinCount = 0;
        int remainingAmount = amount;

        // Step 2 & 3: Iterate through sorted denominations and take as many as possible.
        for (int denomination : denominations) {
            // As long as the current denomination can be used, take it.
            while (remainingAmount >= denomination) {
                remainingAmount -= denomination;
                coinCount++;
            }
            // Optimization: If remainingAmount becomes 0, we've found the solution.
            if (remainingAmount == 0) {
                break;
            }
        }

        // Step 4: Check if the full amount was made.
        // If remainingAmount is not 0, it means the amount cannot be made with the given denominations
        // (or the greedy approach failed because the system is not canonical).
        return remainingAmount == 0 ? coinCount : -1;
    }

    /**
     * **Alternative: Dynamic Programming (Guaranteed Optimal for any coin system)**
     *
     * For arbitrary coin systems (where greedy might fail), Dynamic Programming is the way to go.
     *
     * `dp[i]` represents the minimum number of coins to make change for amount `i`.
     * `dp[0] = 0`
     * `dp[i] = min(dp[i - coin_j] + 1)` for all coin_j <= i
     *
     * This provides a correct solution regardless of the coin system.
     */
    public int findMinCoinsDP(int[] denominations, int amount) {
        if (amount < 0) {
            throw new IllegalArgumentException("Amount cannot be negative.");
        }
        if (amount == 0) {
            return 0;
        }
        if (denominations == null || denominations.length == 0) {
            return -1;
        }

        // Initialize dp array. dp[i] will store the minimum number of coins for amount i.
        // Use amount + 1 as infinity, because the max number of coins can be 'amount' (using 1-cent coins).
        int[] dp = new int[amount + 1];
        Arrays.fill(dp, amount + 1); // Initialize with a value larger than any possible count
        dp[0] = 0; // Base case: 0 coins for 0 amount

        // Iterate through all amounts from 1 to the target amount
        for (int i = 1; i <= amount; i++) {
            // For each amount, consider every denomination
            for (int coin : denominations) {
                // If the current coin can be used to form amount 'i'
                if (coin <= i) {
                    // Update dp[i] if using this coin gives a smaller count
                    // (dp[i - coin] + 1 means min coins for remaining amount + this coin)
                    dp[i] = Math.min(dp[i], dp[i - coin] + 1);
                }
            }
        }

        // If dp[amount] is still amount + 1, it means the amount cannot be formed.
        return dp[amount] > amount ? -1 : dp[amount];
    }

    /**
     * **Time Complexity Analysis (Greedy):**
     * - Sorting denominations: O(D log D), where D is the number of denominations.
     * - Iterating through denominations and using coins: In the worst case, for each denomination,
     *   we might perform `amount / min_denomination` operations. However, a more precise analysis
     *   is that each coin reduces the amount. The total number of coins taken is at most `amount`.
     *   The outer loop runs D times. The inner `while` loop iterations sum up to at most `amount`
     *   over all denominations. So, roughly O(D + Amount).
     * - Overall Time Complexity: O(D log D + Amount). For a fixed set of denominations, this is
     *   effectively O(Amount).
     *
     * **Space Complexity Analysis (Greedy):**
     * - Sorting: O(1) auxiliary space (if `Arrays.sort` modifies in-place) or O(log D) for stack space.
     * - Variables: O(1).
     * - Overall Space Complexity: O(1) auxiliary space.
     *
     * **Time Complexity Analysis (Dynamic Programming):**
     * - Creating and initializing `dp` array: O(Amount).
     * - Nested loops: Outer loop runs `Amount` times, inner loop runs `D` (number of denominations) times.
     * - Overall Time Complexity: O(Amount * D).
     *
     * **Space Complexity Analysis (Dynamic Programming):**
     * - `dp` array: O(Amount) to store minimum coins for each amount up to the target.
     * - Overall Space Complexity: O(Amount).
     */
}

File: leetcode-greedy-algorithms-solutions-advanced/src/problems/coinChange.js
```javascript
/**
 * @fileoverview Implementation of the Coin Change problem using a Greedy approach.
 * This implementation applies the greedy strategy where it is applicable.
 *
 * Problem Description:
 * Given a set of coin denominations and an amount, find the minimum number of
 * coins required to make up that amount.
 *
 * IMPORTANT NOTE ON GREEDY APPROACH FOR COIN CHANGE:
 * The greedy approach (always choosing the largest denomination possible)
 * works ONLY for specific "canonical" coin systems (e.g., US currency: 1, 5, 10, 25 cents;
 * Euro currency: 1, 2, 5, 10, 20, 50 cents, 1, 2 Euro).
 *
 * When the greedy approach *fails*:
 * Consider denominations: [1, 3, 4] and amount: 6
 * Greedy approach:
 * 1. Pick 4 (remaining amount: 2)
 * 2. Pick 1 (remaining amount: 1)
 * 3. Pick 1 (remaining amount: 0)
 * Total coins: 3 ([4, 1, 1])
 *
 * Optimal solution for [1, 3, 4] and amount: 6
 * 1. Pick 3 (remaining amount: 3)
 * 2. Pick 3 (remaining amount: 0)
 * Total coins: 2 ([3, 3])
 *
 * For a general coin system, Dynamic Programming (DP) is required to guarantee
 * an optimal solution. This file implements the greedy strategy for educational
 * purposes, but it's crucial to understand its limitations.
 * The `solutions_comparison/` directory includes a DP solution for the general case.
 *
 * Greedy Choice Property (when it works):
 * For canonical coin systems, taking the largest possible coin at each step
 * ensures that you minimize the total number of coins. The reasoning is usually
 * tied to the specific structure of the coin denominations which prevents smaller
 * coins from "combining" in a way that is more efficient than a larger coin.
 *
 * For example, in US currency:
 * - You'd never use two dimes (20 cents) instead of a quarter (25 cents) to make change for 25 cents.
 * - The sum of any number of smaller coins (excluding the next largest coin) is always less than the next largest coin.
 */

/**
 * Calculates the minimum number of coins needed to make a given amount using a greedy strategy.
 *
 * @param {Array<number>} denominations - An array of available coin denominations.
 *   Must be positive integers.
 * @param {number} amount - The target amount to make change for. Must be a non-negative integer.
 * @returns {Object} An object containing:
 *   - {number} minCoins: The minimum number of coins, or -1 if the amount cannot be made.
 *   - {Array<number>} coinsUsed: An array detailing the count of each coin used, e.g.,
 *     [{ denomination: 25, count: 2 }, { denomination: 10, count: 1 }]
 *   Note: This result is optimal ONLY for canonical coin systems.
 *
 * Time Complexity: O(D log D) for sorting denominations (if not pre-sorted) + O(D) for iteration,
 *                  where D is the number of denominations.
 *                  If denominations are already sorted, it's O(D).
 * Space Complexity: O(D) for storing sorted denominations and coins used.
 */
function coinChangeGreedy(denominations, amount) {
  // Edge cases:
  if (amount < 0) {
    return { minCoins: -1, coinsUsed: [] }; // Cannot make change for negative amount
  }
  if (amount === 0) {
    return { minCoins: 0, coinsUsed: [] }; // 0 coins needed for 0 amount
  }
  if (!denominations || denominations.length === 0) {
    return { minCoins: -1, coinsUsed: [] }; // No denominations available
  }

  // 1. Sort denominations in descending order.
  // This is crucial for the greedy strategy: always try the largest available coin first.
  const sortedDenominations = [...denominations].sort((a, b) => b - a);

  let remainingAmount = amount;
  let totalCoins = 0;
  const coinsUsedMap = new Map(); // To store the count of each denomination used

  // 2. Iterate through sorted denominations.
  for (const coin of sortedDenominations) {
    // If the current coin is greater than 0 (valid denomination)
    // and can be used to reduce the remaining amount.
    if (coin > 0 && remainingAmount >= coin) {
      // Calculate how many times this coin can be used.
      const numCoins = Math.floor(remainingAmount / coin);

      // Add to total coins and update remaining amount.
      totalCoins += numCoins;
      remainingAmount -= numCoins * coin;

      // Record which coins were used.
      coinsUsedMap.set(coin, (coinsUsedMap.get(coin) || 0) + numCoins);
    }

    // If the remaining amount becomes 0, we've found a solution.
    if (remainingAmount === 0) {
      break;
    }
  }

  // 3. Check if the amount was fully made.
  if (remainingAmount !== 0) {
    // If remainingAmount is not 0, it means the amount cannot be made with the given denominations
    // using the greedy strategy. This might happen for non-canonical coin systems.
    return { minCoins: -1, coinsUsed: [] };
  }

  // 4. Format the coinsUsed result.
  const coinsUsed = Array.from(coinsUsedMap.entries()).map(([denomination, count]) => ({
    denomination,
    count
  })).sort((a, b) => b.denomination - a.denomination); // Sort by denomination descending for consistent output

  return { minCoins: totalCoins, coinsUsed };
}

export default coinChangeGreedy;
```